### DISCUSSION TITLE
Waveform

### DISCUSSION THREAD

**Discussion by nicolamas** [vroland/epdiy#340] [Q&A]
Hi everyone, I recently started looking at EPD displays which are really fantastic, by chance I came across the very interesting EPDIY.
I started by looking at the source code but no matter how hard I try I can't understand how the implementation works to use the waveforms, because looking at the header files of the various displays, I thought I was faced with a LUT that accepts the temperature and as output the necessary duration of the +V or -V voltage to go from white to black or from black to white, in order to generate the right gray in a scale of 16.
Instead there are 3-dimensional arrays, in which the content is mostly 0x00
**I'm completely confused!**
I see:
`
typedef struct {
    int phases;
    const uint8_t* luts;
    const int* phase_times;
} EpdWaveformPhases;

typedef struct {
    uint8_t type;
    uint8_t temp_ranges;
    EpdWaveformPhases const** range_data;
} EpdWaveformMode;

typedef struct {
    int min;
    int max;
} EpdWaveformTempInterval;
typedef struct {
    uint8_t num_modes;
    uint8_t num_temp_ranges;
    EpdWaveformMode const** mode_data;
    EpdWaveformTempInterval const* temp_intervals;
} EpdWaveform;
`
Can someone explain to me how it works?
I would be enormously grateful!

**Comment by martinberlin** [comment]
I recommend you reading [this Issue](https://github.com/vroland/epdiy/issues/226) to understand waveforms better.
Most of the waveforms here where done adjusting the timings manually. Adding temperature support is contemplated but I'm not really sure if implemented by @vroland 
I also will take a look at the code in the next weeks so I might tell more. But in that issue I referenced there is more valuable information than my lines here.

**Comment by nicolamas** [comment]
Hello Martin,
I will read the exchange of messages carefully!
Many Thanks,
Nicola

**Reply by martinberlin** [reply]
Additionally I've started writing a Wiki article about this. As soon as I research more, will add code snippets and examples there:

https://github.com/vroland/epdiy/wiki/How-pixels-are-driven-in-a-parallel-epaper-with-epdiy
